### ISSUE TITLE
No permissions available for configuration on first run

### ISSUE THREAD

**Issue by bishwas-medic** [medic/medic-webapp#3251]
On a newly configured V2 instance:

<img width="682" alt="screen shot 2017-03-22 at 4 46 59 pm" src="https://cloud.githubusercontent.com/assets/13689284/24194789/81593556-0f1f-11e7-9f7b-54de7d19650d.png">

**Comment by alxndrsn** [comment]
This prevents us creating any non-admin users.

**Comment by sglangevin** [comment]
Just to be clear, this is for a new instance, not an upgraded instance, right?

**Comment by browndav** [comment]
This is a new instance that had its OS upgraded, and then saw `medic-webapp` upgraded from 0.4.15 to 2.10-alpha. This may be an issue with the upgrade procedure; if it's not something we can immediately solve, I'm happy to recreate this with a non-upgraded image, and then take this one offline to figure out what's going wrong.

**Comment by garethbowen** [comment]
This [migration](https://github.com/medic/medic-api/blob/master/migrations/add-permissions-configuration.js) is supposed to fix this problem. Is it not running or not working?

**Comment by bishwas-medic** [comment]
@garethbowen the meta doc has an entry for `add-permissions-configuration` migration so I think the migration ran but it didn't work.

The same thing happened when trying to config a new instance for a standard project (barefoot.app.medicmobile.org). The Permission page was empty and the list appeared only after uploading an app settings file with permissions defined in it.

**Comment by garethbowen** [comment]
Or maybe it ran but got overwritten by something else...

**Comment by SCdF** [comment]
@bishwas-medic so where is this ticket. FWIW

> The same thing happened when trying to config a new instance for a standard project. The Permission page was empty and the list appeared only after uploading an app settings file with permissions defined in it.

I set up a standard project locally and permissions worked properly. However, "standard" is a project with app_settings… so I'm not sure if that's the same standard that you're talking about.

Do you have a set of steps I can use to replicate this? Like: replicate this DB, then upgrade it to 2.x, then see how there are no permissions; that kind of thing?

**Comment by bishwas-medic** [comment]
@SCdF This happens only in newly created instance and I had to manually add the permission in the app_settings for them to appear on the Permissions page. I have another new instance that I started testing on today which has the same issue. I'll share the URL with you privately.

**Comment by SCdF** [comment]
Talked to @bishwas-medic , by the time he gets them the setup is complete and they are already in this incorrect state.

@abbyad / @henokgetachew / @browndav can any of you shed light on how these get setup?

From talking to Bishwas it sounds like they should just be 2.x running standard, and at least for me that works fine locally.

If someone can give me some reproduction steps, that either start with "replicate this DB" or "have an empty db" so I can reasonably attempt to replicate this, that would be great.

**Comment by henokgetachew** [comment]
There is no AMI that is specific for Standard yet.  Once the instance is up and running, the standard config has to be specifically uploaded to the instance.  I will create a brand new AMI that will support standard out of the box - all the default config options.  But, we would still need to upload the standard settings from medic-projects as the forms and config gets updated.

**Comment by SCdF** [comment]
@henokgetachew I'm looking for reproduction steps. It is currently unclear to me what people are doing to get into this state, since it "works for me". Until I get reproduction steps I'm not sure what else I can do.

**Comment by henokgetachew** [comment]
I mean there is a folder in medic-projects (`default-sms-workflow`) that has all the files and `app_settings.json` that you can just run the upload script over to get your instance up to par.  If these files came by default with a fresh install of the webapp, that could work as well.

**Comment by SCdF** [comment]
@henokgetachew yep done that. Works for me.

Can you please detail the steps that you follow to get an instance installed with standard running 2.x please.

**Comment by henokgetachew** [comment]
@SCdF please use these steps to reproduce.  Go to the garden market and install a fresh medic app from [here](https://staging.dev.medicmobile.org/markets-release-v2/details/medic).  I have confirmed that the instance comes with empty permissions.

**Comment by henokgetachew** [comment]
I have also confirmed that uploading the files in the `default-sms-workflow` solves the issue.  Always doing this step after a clean install could be a workaround for now.  Including these files by default with the webapp would be a better solution however.  That would also solve some of our hierarchy issues we've been running into by using the default contact creation forms that come with the webapp.  Tagging @abbyad too here.

**Comment by henokgetachew** [comment]
@garethbowen, not sure if this is the right question to ask but do the migrations run for a fresh install from garden market?

**Comment by garethbowen** [comment]
@henokgetachew Yeah they do. When API starts it runs each migration that hasn't yet been run on that database. Most migrations won't do anything on first run but it could well be a migration stuffing this up.

**Comment by henokgetachew** [comment]
Alright.  I have just created new AMIs that will have the permissions added by default so that new instances will no longer have this issue.

**Comment by SCdF** [comment]
@henokgetachew so are we saying this ticket is "fixed" then? i.e. no dev required?

**Comment by henokgetachew** [comment]
@SCdF I guess for practical purposes for projects, it won't happen again because the instance will come with a default config.  I don't know if that means it's "fixed" however because you could run into it locally and whatever is making that migration not run would still be there.

**Comment by SCdF** [comment]
OK I'm going to close it for now then. I think if we come across situations where this is an issue we can deal with it then.

**Comment by derickl** [comment]
@SCdF I've just experienced this issue after upgrading from 2.9.1 to 2.10.3 (on a local VM) and uploading the app_settings from lg-uganda.

**Comment by SCdF** [comment]
@derickl that is interesting. FWIW the migration that Gareth mentions near the top of this thread is super old (from 2015!) so you definitely already had it in 2.9.1 already.

Which `app_settings` did you upload?  The one in medic-projects?

**Comment by derickl** [comment]
@SCdF, yep. From medic-projects.

**Comment by SCdF** [comment]
Well that's just the weirdest. So what app settings did you have before that?

And so for clarity (clarify where I'm ??? or wrong):
 - Had 2.9.1 installed on a local medic-os VM, with the ??? app settings and some? / no? data
 - Upgraded (via one click / gardener / etc?) to 2.10.3
 - Ran `./scripts/upload_app_settings.sh lg-uganda/app_settings.json` ??
 - Then logged in and found no permissions where before you had some

Sound right?

**Comment by derickl** [comment]
That's right. 

- Initial install on VM was v2.8.X which was then upgraded to 2.9.1 (a while back)
- I had the default settings.
- Handful of reports on the local VM
- One-click upgrade to 2.10.3
- Ran `upload_app_settings`
- Found a blank permissions page (even after logging out and back in)

**Comment by SCdF** [comment]
By the default settings you don't mean LG-Uganda right? You mean just whatever the app comes with?

It's incredibly weird to me that it would do that, because as I understand it you uploading lg-u would overwrite any incorrect permissions with those ones.

Odd. I will try to reproduce tomorrow

**Comment by derickl** [comment]
Default app settings - whatever the app come with.

**Comment by SCdF** [comment]
@derickl but the permissions looked right when it was on 2.9.1? Or you're not sure / don't remember?

**Comment by derickl** [comment]
@SCdF not sure

**Comment by henokgetachew** [comment]
I'm not completely sure about this but the original instance Bishwas reported about might have first been configured with 2.9 and updated to 2.10 as well.

**Comment by SCdF** [comment]
Steps:
 - Check out 2.9.1 in both webapp and api (based on submodule commit)
 - Have a clean medic db and `grunt dev` into it
 - Run API until migration completion
At this stage if you log in as admin and look in permissions you see an empty permission list.
Then:
 - Check out 2.10.3 in both webapp and api (based on submodule commit)
 - Run `grunt dev` again, pushing the new webapp up
 - Run API until migration completion
At *this* stage permissions looks the same: still empty and wrong
Then:
 - run the upload app settings script to push lg-u's settings up.
As expected (to me), this now *fixes* the permissions page to show the permissions configured in lg-u's app settings.

So there are two things going on here, correct me if I'm wrong @garethbowen:
 - If you use the default app_settings in any version of the app there should be a list of default permissions, and you should not see an empty permissions list
 - Somehow @derickl managed to upload lg-u's settings that *do* contain legit permissions, and still didn't get them. @derickl can you just try again, and doubly-check that the output of the script says that it was successful?

**Comment by derickl** [comment]
@SCdF 

```
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 200 OK
< Server: nginx/1.7.11
< Date: Thu, 06 Apr 2017 12:44:24 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 16
< Connection: keep-alive
< X-Powered-By: Express
< set-cookie: baseURL=/medic/_design/medic/_rewrite; path=/medic/_design/medic/_rewrite
< WWW-Authenticate: Cookie
< X-Frame-Options: SAMEORIGIN
<
* Curl_http_done: called premature == 0
* Connection #0 to host 192.168.182.128 left intact
{"success":true}[upload_app_settings.sh] Completed successfully.
```

**Comment by derickl** [comment]
@SCdF It works on the 2nd upload.

**Comment by SCdF** [comment]
Permissions are working fine.

The real issue is rendering the list:
![image](https://cloud.githubusercontent.com/assets/583851/24902082/faae2ac0-1ea0-11e7-9088-d91a0da54dbb.png)
This is because api has a default config which is being written, and that default config has a duplicate in it.

https://github.com/medic/medic-api/pull/139

@garethbowen take a look

**Comment by garethbowen** [comment]
Wow...

**Comment by sglangevin** [comment]
Any instructions on how to AT? cc @ngaruko

**Comment by ngaruko** [comment]
Any clue how to AT this @garethbowen , @henokgetachew  @anyone?

**Comment by garethbowen** [comment]
AT: run an upgrade from 0.4.x to 2.10.x and make sure you can see a list of permissions when you view the Configuration > Permissions screen.

**Comment by bishwas-medic** [comment]
This issue re-occurred in a freshly upgraded instance. It was upgraded from v0.4.x to v2.12.0. Had to manually replace the permissions in the app_settings to get the list appear on the page.

<img width="593" alt="screen shot 2017-06-28 at 1 53 23 pm" src="https://user-images.githubusercontent.com/13689284/27627053-1ce9ad66-5c0a-11e7-9b01-ebf7e288e034.png">

**Comment by sglangevin** [comment]
@bishwas-medic this error is due to having the same permission listed twice in your `app_settings`. That's what the error message is saying here. We saw this issue during tech lead training and removing the second occurrence of the permission solved the issue. Looks like you have `can_view_analytics` listed twice. You can remove it from the dashboard settings page or re-upload your `app_settings`.

@garethbowen any idea why the `can_view_analytics` permission would get repeated? This is the exact same permission that caused us trouble during the tech lead training. It was not due to an upgrade, so I'm wondering if there may be some other issue here.

**Comment by garethbowen** [comment]
@sglangevin My best guess is that the old configuration @bishwas-medic had already had the duplicate but it didn't cause any issues on the old version of the app. On upgrade the old configuration was retained by design which exposed the issue. The fix for this issue was to fix the default permissions but it won't fix broken permissions on upgrade.

@bishwas-medic Do you have a copy of the pre-upgrade config, or maybe a backup of the instance before upgrading?

**Comment by bishwas-medic** [comment]
@sglangevin yeah, I've already fixed the issue by reuploading the app settings with the changed permissions. The issue is that the duplicate permissions get loaded from the start, before any config  is done. I think that should be addressed.

@garethbowen, to my knowledge, there weren't any permission settings on v0.4.x version of webapp. I'll DM you the old settings file. For backup of the instance before upgrade, Henok might have idea about it. Tagging @henokgetachew here.

**Comment by sglangevin** [comment]
Yeah - the weird thing is that this also happened during the tech lead training on an instance that had working config (including permissions). At some point while one of the teams was working on the config (uploading changes) the analytics permission was duplicated. Seemed odd that it was the exact same permission causing the problem. So perhaps @bishwas-medic is correct and that permission is generated twice by default somehow.

**Comment by garethbowen** [comment]
@SCdF Any thoughts? There are two places in the api code that update the permissions - the config.js which defaults and saves the settings on init, and a migration which updates the settings with default permissions. Maybe the two are conflicting somehow? I'm not sure why we need the migration if the config.js is setting up permissions anyway.

**Comment by SCdF** [comment]
@garethbowen yeah I can't see why the migration is required anymore (with the exception that the defaults code presumed the permissions object exists, but that is a minor change to that code). I presume it existed before the auto-merge. It's interesting that it's the same permission, though I can't see anything in the code that is dodgy about it. 

My suggestion would be to fix it by using a more correct data structure: change permissions to be an object where the property keys are the permission name. Then there can be no duplicates.

**Comment by ngaruko** [comment]
Is this issue related to a new instance @bishwas-medic @sglangevin .Did someone tested it after the fix?

**Comment by bishwas-medic** [comment]
@ngaruko It's been happening in general with v2 instances, not a specific one. I haven't done any testing for this recently.

**Comment by sglangevin** [comment]
@ngaruko judging from the discussion here, we should reopen this issue or create a new issue based on @SCdF's suggestion:

> My suggestion would be to fix it by using a more correct data structure: change permissions to be an object where the property keys are the permission name. Then there can be no duplicates.

**Comment by ngaruko** [comment]
Reopening the issue ATM for a more permanent and stable fix.

**Comment by SCdF** [comment]
I'm closing this in favour of a just raised ticket for changing the data structure: https://github.com/medic/medic-webapp/issues/3762
